test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java

```diff
@@ -0,0 +1,214 @@
+        buffers[0] = firstBuffer;
+        for (int i = 1; i < buffers.length; ++i) {
+            buffers[i] = pageIterator.next().buffer.duplicate();
+        }
+        if (finalLimit != 0) {
+            buffers[buffers.length - 1].limit(finalLimit);
+        }
+
+        return buffers;
+    }
+
+    /**
+     * This method will return an array of {@link ByteBuffer} representing the bytes from the current index
+     * of this buffer through the end. The buffers will be duplicates of the internal buffers, so any
+     * modifications to the markers {@link ByteBuffer#position()}, {@link ByteBuffer#limit()}, etc will not
+     * modify the this class.
+     *
+     * @return the byte buffers
+     */
+    public ByteBuffer[] getPostIndexBuffers() {
+        if (internalIndex == capacity) {
+            return new ByteBuffer[0];
+        }
+        long indexWithOffset = offset + internalIndex;
+
+        int pageIndex = pageIndex(indexWithOffset);
+        int indexInPage = indexInPage(indexWithOffset);
+
+        ByteBuffer[] buffers = new ByteBuffer[pages.size() - pageIndex];
+        Iterator<Page> pageIterator = pages.descendingIterator();
+        for (int i = buffers.length - 1; i > 0; --i) {
+            buffers[i] = pageIterator.next().buffer.duplicate();
+        }
+        ByteBuffer firstPostIndexBuffer = pageIterator.next().buffer.duplicate();
+        firstPostIndexBuffer.position(firstPostIndexBuffer.position() + indexInPage);
+        buffers[0] = firstPostIndexBuffer;
+
+        return buffers;
+    }
+
+    public void incrementIndex(long delta) {
+        if (delta < 0) {
+            throw new IllegalArgumentException("Cannot increment an index with a negative delta [" + delta + "]");
+        }
+
+        long newIndex = delta + internalIndex;
+        if (newIndex > capacity) {
+            throw new IllegalArgumentException("Cannot increment an index [" + internalIndex + "] with a delta [" + delta +
+                "] that will result in a new index [" + newIndex + "] that is greater than the capacity [" + capacity + "].");
+        }
+        internalIndex = newIndex;
+    }
+
+    public long getIndex() {
+        return internalIndex;
+    }
+
+    public long getCapacity() {
+        return capacity;
+    }
+
+    public long getRemaining() {
+        return capacity - internalIndex;
+    }
+
+    private int numPages(long capacity) {
+        final long numPages = (capacity + pageMask) >>> pageShift;
+        if (numPages > Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("pageSize=" + (pageMask + 1) + " is too small for such as capacity: " + capacity);
+        }
+        return (int) numPages;
+    }
+
+    private int pageIndex(long index) {
+        return (int) (index >>> pageShift);
+    }
+
+    private int indexInPage(long index) {
+        return (int) (index & pageMask);
+    }
+
+    private static class Page implements Releasable {
```

Comment:
can we maybe remove this class all-together since we don't use it and can we introduce it once it's needed?